File: backend/internal/models/lesson_plan_section_rewrite.go
package models

// lesson_plan_section_rewrite.go — 教案目录与段落AI修改的数据协议。
//
// 设计原则：
//   1. 教案正文仍然以lesson_plans.content_markdown作为唯一事实源。
//   2. 目录和段落结构由代码确定性解析，不写入数据库，不调用AI生成目录。
//   3. AI修改采用“生成预览”和“确认应用”两阶段，生成预览绝不修改正文。
//   4. 确认应用必须同时校验教案版本与段落正文哈希，防止旧页面覆盖新正文。
//   5. 只替换标题下方的直属正文，不修改标题行，也不自动覆盖子标题结构。

// LessonPlanSectionLocator 是浏览器定位教案段落时提交的最小定位信息。
//
// HeadingText保存Markdown中的原始标题行去除首尾空白后的文本，
// 例如“## 教学目标”“一、教学过程”“**板书设计**”。
//
// Occurrence用于区分正文中重复出现的相同标题，从1开始计数。
// 后端始终重新解析数据库正式正文，不能直接相信浏览器提交的段落内容。
type LessonPlanSectionLocator struct {
	HeadingText string `json:"heading_text"`
	Occurrence  int    `json:"occurrence"`
}

// LessonPlanDocumentSection 是确定性解析得到的教案目录节点和可编辑正文范围。
//
// StartOffset、ContentStartOffset和EndOffset都是UTF-8字符串的字节偏移，
// 仅用于后端内部精确切片。浏览器不应自行提交这些偏移作为授权依据。
//
// BodyMarkdown只包含当前标题到下一个标题之间的直属正文，
// 不包含标题行，也不吞并后续子标题，避免一次修改意外覆盖整个章节结构。
type LessonPlanDocumentSection struct {
	ID                 string                   `json:"id"`
	Title              string                   `json:"title"`
	HeadingText        string                   `json:"heading_text"`
	Level              int                      `json:"level"`
	HeadingPath        []string                 `json:"heading_path"`
	Occurrence         int                      `json:"occurrence"`
	StartOffset        int                      `json:"-"`
	ContentStartOffset int                      `json:"-"`
	EndOffset          int                      `json:"-"`
	BodyMarkdown       string                   `json:"body_markdown"`
	SectionHash        string                   `json:"section_hash"`
	Locator            LessonPlanSectionLocator `json:"locator"`
}

// GenerateLessonPlanSectionRewriteRequest 请求AI生成某个教案段落的修改预览。
//
// BaseVersion必须等于数据库当前教案版本；版本不一致时必须刷新后重试。
// Instruction是老师本次修改要求，不能包含身份、学校或教育域等授权字段。
type GenerateLessonPlanSectionRewriteRequest struct {
	BaseVersion int                      `json:"base_version"`
	Locator     LessonPlanSectionLocator `json:"locator"`
	Instruction string                   `json:"instruction"`
}

// LessonPlanSectionRewritePreview 是AI生成完成后的预览结果。
//
// ReplacementMarkdown只包含可替换进当前标题下方的正文，
// 不含标题行。老师确认前不会写入数据库。
type LessonPlanSectionRewritePreview struct {
	BaseVersion        int                       `json:"base_version"`
	Section            LessonPlanDocumentSection `json:"section"`
	ReplacementMarkdown string                    `json:"replacement_markdown"`
}

// ApplyLessonPlanSectionRewriteRequest 是老师确认采用AI建议时的请求。
//
// SectionHash必须使用生成预览接口返回的服务端哈希，
// 不能由浏览器使用当前显示文本重新猜测。
type ApplyLessonPlanSectionRewriteRequest struct {
	BaseVersion        int                      `json:"base_version"`
	Locator            LessonPlanSectionLocator `json:"locator"`
	SectionHash        string                   `json:"section_hash"`
	ReplacementMarkdown string                   `json:"replacement_markdown"`
}

// LessonPlanSectionRewriteApplyResponse 是段落修改原子写入后的结果。
//
// Changed=false表示替换后的完整正文与数据库当前正文完全相同，
// 此时不会新增版本快照，也不会递增版本号。
type LessonPlanSectionRewriteApplyResponse struct {
	Changed         bool   `json:"changed"`
	CurrentVersion  int    `json:"current_version"`
	ContentMarkdown string `json:"content_markdown"`
}

File: frontend/src/api/lesson-plan-section-rewrite.ts
/**
 * lesson-plan-section-rewrite.ts — 教案目录段落AI修改API。
 *
 * 两阶段协议：
 *   1. section-rewrite：流式生成预览，不写数据库。
 *   2. section-rewrite/apply：老师确认后原子应用。
 *
 * 应用接口会由后端再次校验作者、教案状态、版本和段落哈希。
 */

import apiClient from './client'

export interface LessonPlanSectionLocator {
  heading_text: string
  occurrence: number
}

export interface LessonPlanServerSection {
  id: string
  title: string
  heading_text: string
  level: number
  heading_path: string[]
  occurrence: number
  body_markdown: string
  section_hash: string
  locator: LessonPlanSectionLocator
}

export interface GenerateLessonPlanSectionRewriteRequest {
  base_version: number
  locator: LessonPlanSectionLocator
  instruction: string
}

export interface LessonPlanSectionRewritePreview {
  base_version: number
  section: LessonPlanServerSection
  replacement_markdown: string
}

export interface ApplyLessonPlanSectionRewriteRequest {
  base_version: number
  locator: LessonPlanSectionLocator
  section_hash: string
  replacement_markdown: string
}

export interface LessonPlanSectionRewriteApplyResponse {
  changed: boolean
  current_version: number
  content_markdown: string
}

export interface LessonPlanSectionRewriteHandlers {
  onConnected?: () => void
  onChunk?: (chunk: string) => void
  onDone?: (preview: LessonPlanSectionRewritePreview) => void
  onError?: (message: string, code?: string) => void
}

export interface LessonPlanSectionRewriteConnection {
  close: () => void
}

interface SectionRewriteSSEError {
  code?: string
  message?: string
}

interface SectionRewriteSSEDone {
  preview?: LessonPlanSectionRewritePreview
}

/**
 * 流式生成教案段落修改预览。
 *
 * 该连接是一次性的，不执行自动重连。
 * 中断后由老师点击“重新生成”，避免同一修改要求重复扣费。
 */
export function generateLessonPlanSectionRewrite(
  planID: string,
  request: GenerateLessonPlanSectionRewriteRequest,
  handlers: LessonPlanSectionRewriteHandlers,
): LessonPlanSectionRewriteConnection {
  const controller = new AbortController()
  const token = localStorage.getItem('token') || ''

  let closed = false
  let terminalEventReceived = false

  const emitError = (message: string, code?: string) => {
    if (closed) return
    terminalEventReceived = true
    handlers.onError?.(message, code)
  }

  void fetch(
    `/api/v1/lesson-plans/plans/${encodeURIComponent(planID)}/section-rewrite`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(request),
      signal: controller.signal,
    },
  )
    .then(async response => {
      if (!response.ok) {
        const payload = await response
          .json()
          .catch(() => null) as { message?: string } | null

        emitError(
          payload?.message || `请求失败：HTTP ${response.status}`,
          `http_${response.status}`,
        )
        return
      }

      if (!response.body) {
        emitError('当前浏览器不支持流式响应', 'stream_unsupported')
        return
      }

      const reader = response.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''

      try {
        while (!closed) {
          const { done, value } = await reader.read()
          if (done) break

          buffer += decoder.decode(value, { stream: true })
          buffer = buffer.replace(/\r\n/g, '\n')

          const blocks = buffer.split('\n\n')
          buffer = blocks.pop() || ''

          for (const block of blocks) {
            if (!block.trim()) continue

            let eventType = ''
            const dataLines: string[] = []

            for (const line of block.split('\n')) {
              if (line.startsWith('event:')) {
                eventType = line.slice(6).trim()
              } else if (line.startsWith('data:')) {
                dataLines.push(line.slice(5).trimStart())
              }
            }

            if (!eventType || dataLines.length === 0) continue

            let payload: unknown
            try {
              payload = JSON.parse(dataLines.join('\n'))
            } catch {
              continue
            }

            if (eventType === 'connected') {
              handlers.onConnected?.()
              continue
            }

            if (eventType === 'chunk') {
              const data = payload as { chunk?: string }
              if (data.chunk) handlers.onChunk?.(data.chunk)
              continue
            }

            if (eventType === 'error') {
              const data = payload as SectionRewriteSSEError
              emitError(
                data.message || 'AI修改建议生成失败',
                data.code,
              )
              closed = true
              break
            }

            if (eventType === 'done') {
              const data = payload as SectionRewriteSSEDone
              if (!data.preview) {
                emitError('AI返回的修改预览不完整', 'invalid_preview')
              } else {
                terminalEventReceived = true
                handlers.onDone?.(data.preview)
              }
              closed = true
              break
            }
          }
        }
      } finally {
        await reader.cancel().catch(() => undefined)
      }

      if (!closed && !terminalEventReceived) {
        emitError('流式连接提前结束，请重新生成', 'stream_interrupted')
      }
    })
    .catch(error => {
      if (closed || error?.name === 'AbortError') return

      const message = error instanceof Error
        ? error.message
        : '网络连接失败'

      emitError(message, 'network_error')
    })

  return {
    close: () => {
      if (closed) return
      closed = true
      controller.abort()
    },
  }
}

/** 原子应用老师确认的段落修改结果。 */
export async function applyLessonPlanSectionRewrite(
  planID: string,
  request: ApplyLessonPlanSectionRewriteRequest,
): Promise<LessonPlanSectionRewriteApplyResponse> {
  const response = await apiClient.post(
    `/lesson-plans/plans/${planID}/section-rewrite/apply`,
    request,
  )

  return response.data.data as LessonPlanSectionRewriteApplyResponse
}
